MINOR: fix Streams version-probing system test

mjsax opened the pull request:
With KAFKA-8284, the `AbstractCoordinator` does not allow to suppress sending `LeaveGroupRequest` on `Consumer#close()` any longer. Hence, we get a different rebalance pattern.

This implies, that the last rebalance, does not do any version probing any longer. When the last instance is bounced, an immediate rebalance "stabilized the group" and prepares it to upgrade to the new metadata version, because all members are on old version but support new version already. On restart of the last instance, we immediately switch to the new version instead of doing a version probing.

mjsax committed:
MINOR: fix Streams version-probing system test

mjsax commented:
Retest this please

mjsax commented:
Call for review @guozhangwang @bbejeck @vvcephei @ableegoldman @cadonna @abbccdda

mjsax commented:
System test passed: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2611/

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java:
```diff
@@ -90,16 +89,13 @@ public static void main(final String[] args) throws Exception {
         final KafkaStreams streams = new KafkaStreams(builder.build(), config, kafkaClientSupplier);
         streams.start();
 
-        Runtime.getRuntime().addShutdownHook(new Thread() {
```
Just Java8 cleanup.

mjsax commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -348,7 +349,7 @@ def update_leader(self):
             for p in self.processors:
                 found = list(p.node.account.ssh_capture("grep \"Finished assignment for group\" %s" % p.LOG_FILE, allow_fail=True))
-                if len(found) == self.leader_counter[p] + 1:
+                if len(found) >= self.leader_counter[p] + 1:
```
This is one fix. If rebalance happens quickly, we might see this message for the second rebalance early. Cf. #6763

mjsax replied:
After https://github.com/apache/kafka/pull/6779 is merged, I can rebase this PR -- we still want this fix I guess :)

mjsax commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -564,28 +563,31 @@ def do_rolling_bounce(self, processor, counter, current_generation):
-                    leader_monitor.wait_until("Received a future (version probing) subscription (version: 5). Sending empty assignment back (with supported version 4).",
-                                              timeout_sec=60,
-                                              err_msg="Could not detect 'version probing' attempt at leader " + str(self.leader.node.account))
+                    if counter < 3:
```
This is the actual fix (similar below).

mjsax commented:
Call for review @guozhangwang @bbejeck @vvcephei @ableegoldman @cadonna @abbccdda

vvcephei reviewed (approved):
Ah. Thanks for figuring this out!

guozhangwang commented:
I think we should consider whether streams/consumer should still maintain the `leaveOnshutdown` flag or not, and depending on that this PR should be refactored.

guozhangwang commented:
cc @abbccdda

guozhangwang commented:
@mjsax could you rebase this one given the other PR has been merged now?

abbccdda commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -564,28 +563,31 @@ def do_rolling_bounce(self, processor, counter, current_generation):
-                    leader_monitor.wait_until("Received a future (version probing) subscription (version: 5). Sending empty assignment back (with supported version 4).",
-                                              timeout_sec=60,
-                                              err_msg="Could not detect 'version probing' attempt at leader " + str(self.leader.node.account))
+                    if counter < 3:
```
nit: do we want to define `3` instead of leaving it as magic number?

mjsax replied:
With the latest fixes, I think we actually don't need this any longer. We only needed if, if we get immediate rebalance. With disabling leave group request, this this should not be required any longer.

abbccdda reviewed (approved):
LGTM with minor comment

mjsax commented:
Rebased and updated this PR. Version probing system test passed locally. Also triggered https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2629/

If no objection, I'll merge this after Jenkins and system test build passed.

guozhangwang reviewed (approved):
@mjsax feel free to merge after green builds.

mjsax commented:
System test passed. Jenkins failed with env issue.

Retest this please.

ijuma commented:
@mjsax seems ready to merge.

mjsax commented:
Merged to `trunk` and cherry-picked to `2.3`.
